File: backend/app/routers/customers.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional
from uuid import UUID
from app.dependencies import get_db, require_roles
from app.models.user import UserRole
from app.models.customer import Customer
from app.schemas.customer import CustomerCreate, CustomerUpdate, CustomerOut, CustomerWithOrders

router = APIRouter(prefix="/customers", tags=["customers"])

_view   = require_roles(UserRole.admin, UserRole.manager, UserRole.kassir)
_edit   = require_roles(UserRole.admin, UserRole.manager)
_delete = require_roles(UserRole.admin)


@router.get("", response_model=dict)
def list_customers(
    search: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=500),
    db: Session = Depends(get_db),
    _=Depends(_view),
):
    query = db.query(Customer)
    if search:
        query = query.filter(
            Customer.full_name.ilike(f"%{search}%") |
            Customer.email.ilike(f"%{search}%") |
            Customer.phone.ilike(f"%{search}%")
        )
    total = query.count()
    customers = query.order_by(Customer.created_at.desc()).offset(skip).limit(limit).all()
    return {
        "total": total,
        "skip": skip,
        "limit": limit,
        "items": [CustomerOut.model_validate(c) for c in customers],
    }


@router.post("", response_model=CustomerOut, status_code=201)
def create_customer(data: CustomerCreate, db: Session = Depends(get_db), _=Depends(_edit)):
    customer = Customer(**data.model_dump())
    db.add(customer)
    db.commit()
    db.refresh(customer)
    return customer


@router.get("/{customer_id}", response_model=CustomerWithOrders)
def get_customer(customer_id: UUID, db: Session = Depends(get_db), _=Depends(_view)):
    customer = db.query(Customer).filter(Customer.id == customer_id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    return customer


@router.put("/{customer_id}", response_model=CustomerOut)
def update_customer(customer_id: UUID, data: CustomerUpdate, db: Session = Depends(get_db), _=Depends(_edit)):
    customer = db.query(Customer).filter(Customer.id == customer_id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(customer, key, value)
    db.commit()
    db.refresh(customer)
    return customer


@router.delete("/{customer_id}", status_code=204)
def delete_customer(customer_id: UUID, db: Session = Depends(get_db), _=Depends(_delete)):
    customer = db.query(Customer).filter(Customer.id == customer_id).first()
    if not customer:
        raise HTTPException(status_code=404, detail="Customer not found")
    db.delete(customer)
    db.commit()

File: frontend/src/stores/customers.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/api'

export const useCustomersStore = defineStore('customers', () => {
  const customers = ref([])
  const total = ref(0)
  const loading = ref(false)

  async function fetchCustomers(params = {}) {
    loading.value = true
    try {
      const res = await api.get('/customers', { params })
      customers.value = res.data.items
      total.value = res.data.total
    } finally {
      loading.value = false
    }
  }

  async function getCustomer(id) {
    const res = await api.get(`/customers/${id}`)
    return res.data
  }

  async function createCustomer(data) {
    const res = await api.post('/customers', data)
    return res.data
  }

  async function updateCustomer(id, data) {
    const res = await api.put(`/customers/${id}`, data)
    return res.data
  }

  async function deleteCustomer(id) {
    await api.delete(`/customers/${id}`)
  }

  return { customers, total, loading, fetchCustomers, getCustomer, createCustomer, updateCustomer, deleteCustomer }
})
